funcx_endpoint/funcx_endpoint/endpoint/rabbit_mq/result_queue_subscriber.py

```diff
@@ -0,0 +1,316 @@
+import logging
+import multiprocessing
+from typing import Tuple
+
+import pika
+
+logger = logging.getLogger(__name__)
+
+
+class ResultQueueSubscriber:
+    """The ResultQueueSubscriber is a direct rabbitMQ pipe subscriber that uses
+    the SelectConnection adaptor to enable performance consumption of messages
+    from the service
+
+    """
+
+    def __init__(
+        self,
+        pika_conn_params,
+        exchange="results",
+        exchange_type="topic",
+    ):
+
+        self.queue_name = "results"
+        self.routing_key = "*.results"
+        self.params = pika_conn_params
+        self.exchange = exchange
+        self.exchange_type = exchange_type
+        self._is_connected = False
+        logger.debug("Init done")
+
+    def connect(self):
+        logger.info("Connecting to %s", self.params)
+        return pika.SelectConnection(
+            pika.URLParameters("amqp://guest:guest@localhost:5672/%2F"),  # self.params,
+            on_open_callback=self._on_connection_open,
+            on_close_callback=self._on_connection_closed,
+            on_open_error_callback=self._on_open_failed,
+        )
+
+    def _on_connection_open(self, unused_connection):
+        logger.debug("Connection opened")
+        self.open_channel()
+
+    def _on_open_failed(self, *args):
+        logger.warning(f"Connection failed to open : {args}")
+
+    def _on_connection_closed(self, connection, exception):
```

Comment:
Seeing this method and how it's used to construct the `SelectConnection` is driving me more strongly to think that we should isolate creation and management of the connection objects into a dedicated set of helpers. These are details we should be working out how to share between the different connection classes.

For example,
```python
def on_connection_closed(connection: pika.Connection, exception: Exception) -> None:
    if isinstance(connection, pika.SelectConnection):
        _on_select_connection_closed(connection, exception)
    elif isinstance(connection, pika.BlockingConnection):
        _on_blocking_connection_closed(connection, exception)
    ...
```